src/libraries/Common/src/System/Net/Http/Http3/QPack/QPackEncoder.cs

```diff
@@ -0,0 +1,424 @@
+        public static bool EncodeValueString(ReadOnlySpan<string> values, string separator, Span<byte> buffer, out int length)
+        {
+            if (values.Length == 1)
+            {
+                return EncodeValueString(values[0], buffer, out length);
+            }
+
+            if (values.Length == 0)
+            {
+                return EncodeValueString(string.Empty, buffer, out length);
+            }
+
+            if (buffer.Length > 0)
+            {
+                int valueLength = separator.Length * (values.Length - 1);
+                for (int i = 0; i < values.Length; ++i)
+                {
+                    valueLength += values[i].Length;
+                }
+
+                buffer[0] = 0;
+                if (IntegerEncoder.Encode(valueLength, 7, buffer, out int nameLength))
+                {
+                    buffer = buffer.Slice(nameLength);
+                    if (buffer.Length >= valueLength)
+                    {
+                        string value = values[0];
+                        EncodeValueStringPart(value, buffer);
+                        buffer = buffer.Slice(value.Length);
+
+                        for (int i = 1; i < values.Length; ++i)
+                        {
+                            EncodeValueStringPart(separator, buffer);
+                            buffer = buffer.Slice(separator.Length);
+
+                            value = values[i];
+                            EncodeValueStringPart(value, buffer);
+                            buffer = buffer.Slice(value.Length);
+                        }
+
+                        length = nameLength + valueLength;
+                        return true;
+                    }
+                }
+            }
+
+            length = 0;
+            return false;
+        }
+
+        private static void EncodeValueStringPart(string s, Span<byte> buffer)
+        {
+            Debug.Assert(buffer.Length >= s.Length);
+
+            for (int i = 0; i < s.Length; ++i)
+            {
+                char ch = s[i];
+
+                if (ch > 127)
+                {
+                    throw new QPackEncodingException("ASCII header value.");
+                }
+
+                buffer[i] = (byte)ch;
+            }
+        }
+
+        private static bool EncodeNameString(string s, Span<byte> buffer, out int length)
+        {
+            const int toLowerMask = 0x20;
+
+            if (buffer.Length != 0)
+            {
+                buffer[0] = 0x30;
+
+                if (IntegerEncoder.Encode(s.Length, 3, buffer, out int nameLength))
+                {
+                    buffer = buffer.Slice(nameLength);
+
+                    if (buffer.Length >= s.Length)
+                    {
+                        for (int i = 0; i < s.Length; ++i)
+                        {
+                            int ch = s[i];
+                            Debug.Assert(ch <= 127, "QPack header names must be in ASCII.");
```

Comment:
Upon discovering non-ASCII, we assert when encoding the name but throw when encoding the value?  Did we already validate the name earlier?

Reply:
I believe I did this because HttpHeadersCollection performs this check already, but can't recall 100%. Will double check.